### PR TITLE
Bindings python

### PR DESCRIPTION
**Pull request by ackRow**
## Features

* New python interfaces based on objects rather than functions
* Unit tests for the python lib
* CI: automatic building, testing and publishing

## Technical choices

* Compatibility with manylinux glibc >= 2.17 and Python >= 3.7
[Manylinux](https://github.com/pypa/manylinux#manylinux)
[Abi3](https://pyo3.rs/v0.17.3/features.html)

## Limitations

* No build pipeline for Windows/MacOS yet

* Doc string exported but not the functions signature

### CODE REVIEW

File: src/interfaces/pyo3/py_abe_policy.rs [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+            Ok(att) => att,
+            Err(e) => return Err(PyException::new_err(e.to_string())),
+        };
+        Ok(Attribute { inner: attribute })
+    }
+}
+
+/// Defines a policy axis by its name and its underlying attribute names.
+///
+/// If `hierarchical` is set to `true`, we assume a lexicographical order based
```
Comment (tbrezot) [inline src/interfaces/pyo3/py_abe_policy.rs:45]: We do not have any `hierarchical` variable here.

```suggestion
/// If the axis is defined as hierarchical, we assume a lexicographical order
```

File: src/interfaces/pyo3/py_abe_policy.rs [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+            inner: self.inner.clone(),
+        }
+    }
+
+    /// JSON serialization
+    pub fn to_json(&self) -> PyResult<String> {
+        match serde_json::to_string(&self.inner) {
+            Ok(res) => Ok(res),
+            Err(e) => Err(PyException::new_err(e.to_string())),
+        }
```
Comment (tbrezot) [inline src/interfaces/pyo3/py_abe_policy.rs:158]:
```suggestion
        serde_json::to_string(&self.inner).map_err(|e| PyException::new_err(e.to_string()))
```

File: src/interfaces/pyo3/py_cover_crypt.rs [hunk 3]
```diff
@@ -0,0 +1,404 @@
...
+// https://pyo3.rs/v0.16.2/class.html
+
+#[pyclass]
+pub struct MasterSecretKey {
+    inner: MasterSecretKeyRust,
+}
+
+#[pymethods]
+impl MasterSecretKey {
+    /// Convert to bytes for sending/saving the key
```
Comment (tbrezot) [inline src/interfaces/pyo3/py_cover_crypt.rs:28]: The user will define his needs by himself, no need to mention why one would want to convert to bytes here :)

File: src/interfaces/pyo3/py_cover_crypt.rs [hunk 4]
```diff
@@ -0,0 +1,404 @@
...
+        match self.inner.try_to_bytes() {
+            Ok(res) => Ok(res),
+            Err(e) => Err(PyException::new_err(e.to_string())),
+        }
+    }
+
+    /// Read key from bytes
+    #[classmethod]
+    pub fn from_bytes(_cls: &PyType, key_bytes: Vec<u8>) -> PyResult<Self> {
+        match PublicKeyRust::try_from_bytes(&key_bytes) {
```
Comment (tbrezot) [inline src/interfaces/pyo3/py_cover_crypt.rs:64]: We may not want to take ownership here, though I don't really know what it means in python.

```suggestion
    pub fn from_bytes(_cls: &PyType, key_bytes: &[u8]) -> PyResult<Self> {
        match PublicKeyRust::try_from_bytes(key_bytes) {
```

Do this everywhere you can

File: src/interfaces/pyo3/py_cover_crypt.rs [hunk 5]
```diff
@@ -0,0 +1,404 @@
...
+#[pyclass]
+pub struct SymmetricKey {
+    inner: SymmetricKeyRust,
+}
+
+#[pymethods]
+impl SymmetricKey {
+    /// Convert to bytes for sending/saving the key
+    pub fn to_bytes(&self) -> PyResult<Vec<u8>> {
+        Ok(self.inner.to_bytes().to_vec())
```
Comment (tbrezot) [inline src/interfaces/pyo3/py_cover_crypt.rs:105]: This avoids one copy:
```suggestion
        Ok(self.inner.as_bytes().to_vec())
```
(+ add `use cosmian_crypto_core::symmetric_crypto::SymKey`)

File: src/interfaces/pyo3/py_cover_crypt.rs [hunk 6]
```diff
@@ -0,0 +1,404 @@
...
+    ///
+    /// A new user secret key does NOT include to old (i.e. rotated) partitions
+    ///
+    /// - `msk`                 : master secret key
+    /// - `access_policy_str`   : user access policy
+    /// - `policy`              : global policy
+    pub fn generate_user_secret_key(
+        &self,
+        msk: &MasterSecretKey,
+        access_policy_str: String,
```
Comment (tbrezot) [inline src/interfaces/pyo3/py_cover_crypt.rs:177]: Maybe use `&str`

Do this everywhere you can

File: src/interfaces/pyo3/py_cover_crypt.rs [hunk 7]
```diff
@@ -0,0 +1,404 @@
...
+        {
+            Ok(usk) => Ok(UserSecretKey { inner: usk }),
+            Err(e) => Err(PyException::new_err(e.to_string())),
+        }
+    }
+
+    /// Refresh the user key according to the given master key and access policy.
+    ///
+    /// The user key will be granted access to the current partitions, as determined by its access policy.
+    /// If preserve_old_partitions_access is set, the user access to rotated partitions will be preserved
```
Comment (tbrezot) [inline src/interfaces/pyo3/py_cover_crypt.rs:195]:
```suggestion
    /// If `preserve_old_partitions_access` is set, the user access to rotated partitions will be preserved
```

File: src/interfaces/pyo3/py_cover_crypt.rs [hunk 8]
```diff
@@ -0,0 +1,404 @@
...
+        ) {
+            Ok(()) => Ok(()),
+            Err(e) => Err(PyException::new_err(e.to_string())),
+        }
+    }
+
+    /// Encrypt data symmetrically in a block.
+    ///
+    /// - `symmetric_key`       : symmetric key
+    /// - `plaintext_bytes`     : plaintext to encrypt
```
Comment (tbrezot) [inline src/interfaces/pyo3/py_cover_crypt.rs:228]: Doc does not match the signature

File: src/interfaces/pyo3/py_cover_crypt.rs [hunk 9]
```diff
@@ -0,0 +1,404 @@
...
+}
+
+#[pymethods]
+impl MasterSecretKey {
+    /// Convert to bytes for sending/saving the key
+    pub fn to_bytes(&self) -> PyResult<Vec<u8>> {
+        match self.inner.try_to_bytes() {
+            Ok(res) => Ok(res),
+            Err(e) => Err(PyException::new_err(e.to_string())),
+        }
```
Comment (tbrezot) [inline src/interfaces/pyo3/py_cover_crypt.rs:33]: Do the same everywhere you can.

```suggestion
        self.inner.try_to_bytes().map_err(PyErr::from)
```

File: src/interfaces/pyo3/py_cover_crypt.rs [hunk 10]
```diff
@@ -0,0 +1,404 @@
...
+    }
+
+    /// Read key from bytes
+    #[classmethod]
+    pub fn from_bytes(_cls: &PyType, key_bytes: Vec<u8>) -> PyResult<Self> {
+        match MasterSecretKeyRust::try_from_bytes(&key_bytes) {
+            Ok(key) => Ok(MasterSecretKey { inner: key }),
+            Err(e) => Err(PyException::new_err(e.to_string())),
```
Comment (tbrezot) [inline src/interfaces/pyo3/py_cover_crypt.rs:41]: Do this everywhere you can
```suggestion
            Err(e) => Err(e.into())),
```

File: src/interfaces/pyo3/py_cover_crypt.rs [hunk 11]
```diff
@@ -0,0 +1,404 @@
...
+    /// - `symmetric_key`       : symmetric key
+    /// - `plaintext_bytes`     : plaintext to encrypt
+    /// - `authenticated_data`  : associated data to be passed to the DEM scheme
+    pub fn encrypt_symmetric_block(
+        &self,
+        symmetric_key: &SymmetricKey,
+        plaintext: Vec<u8>,
+        authenticated_data: Option<Vec<u8>>,
+    ) -> PyResult<Vec<u8>> {
+        Ok(self.inner.encrypt(
```
Comment (tbrezot) [inline src/interfaces/pyo3/py_cover_crypt.rs:236]: Maybe use a `map_err()` here too

**Review by tbrezot — CHANGES_REQUESTED** [review]
Very nice! Some changes to be made on error conversion.

Also, you may want to implement your wrappers that way:
``` rust
struct Attribute(AttributeRust);
```
This is less noisy then:
``` rust
pub fn new(axis: &str, name: &str) -> Self {
    Self(AttributeRust::new(axis, name))
}

File: src/interfaces/pyo3/py_abe_policy.rs [hunk 12]
```diff
@@ -0,0 +1,168 @@
...
+impl PolicyAxis {
+    /// Generates a new policy axis with the given name and attribute names.
+    /// A hierarchical axis enforces order between its attributes.
+    ///
+    /// - `name`        : axis name
+    /// - `attributes`  : name of the attributes on this axis
+    /// - `hierarchical`: set the axis to be hierarchical
+    #[new]
+    fn new(name: &str, attributes: Vec<&str>, hierarchical: bool) -> Self {
+        PolicyAxis {
```
Comment (tbrezot) [inline src/interfaces/pyo3/py_abe_policy.rs:62]: Think about using `Self` where you can
```suggestion
        Self {
```

File: .gitlab-ci.yml [hunk 13]
```diff
@@ -67,7 +68,7 @@ is_publishable:
 .base_compile: &base_compile
   stage: build
   cache:
-    key: "${CI_COMMIT_REF_SLUG}"
```
[the remark this answers is not in the document]
Comment (ackRow) [reply .gitlab-ci.yml:70]: I used pre-commit automatic formatting

File: src/interfaces/pyo3/py_abe_policy.rs [hunk 14]
```diff
@@ -0,0 +1,168 @@
+use abe_policy::{Attribute as AttributeRust, Policy as PolicyRust, PolicyAxis as PolicyAxisRust};
+use pyo3::{exceptions::PyException, exceptions::PyTypeError, prelude::*, types::PyType};
+
+// Pyo3 doc on classes
+// https://pyo3.rs/v0.16.2/class.html
+
+/// An attribute in a policy group is characterized by the axis policy name
```
[the remark this answers is not in the document]
Comment (ackRow) [reply src/interfaces/pyo3/py_abe_policy.rs:7]: yes !

File: src/interfaces/pyo3/py_cover_crypt.rs (hunk 4 above)
Comment (ackRow) [reply src/interfaces/pyo3/py_cover_crypt.rs:59]: Seems not possible: TypeError: argument 'key_bytes': 'list' object cannot be converted to 'PyBytes'

File: Cargo.toml [hunk 15]
```diff
@@ -44,7 +44,8 @@ zeroize = "1.5"
 # Optional ones
 js-sys = { version = "0.3", optional = true }
 lazy_static = { version = "1.4", optional = true }
-pyo3 = { version = "0.16.3", features = ["extension-module"], optional = true }
+# Build python package for version 3.7 and above: https://pyo3.rs/v0.17.3/features.html
```
Comment (Manuthor) [inline Cargo.toml:47]: no specific comment for this line but rather on the file Cargo.toml:
- we must bump the version to 8.0.0 du to:
```
Description:
A publicly-visible struct cannot be imported by its prior path. A `pub use` may have been removed, or the struct itself may have been renamed or removed entirely.
        ref: https://doc.rust-lang.org/cargo/reference/semver.html#item-remove
       impl: https://github.com/obi1kenobi/cargo-semver-check/tree/v0.12.1/src/queries/struct_missing.ron

Failed in:
  struct cosmian_cover_crypt::api::ClearTextHeader, previously in file /home/manu/.cargo/registry/src/github.com-1ecc6299db9ec823/cosmian_cover_crypt-7.0.1/src/api.rs:339
  struct cosmian_cover_crypt::ClearTextHeader, previously in file /home/manu/.cargo/registry/src/github.com-1ecc6299db9ec823/cosmian_cover_crypt-7.0.1/src/api.rs:339
```
- why do we use the `develop` branch of `cosmian_crypto_core`? (crate publishing will fail outside a version reference)
- `criterion` version can be upgraded to `0.4`
- we could do some reordering as follow (if it is OK for you):
```
[package]
name = "cosmian_cover_crypt"
version = "8.0.0"
authors = [
  "Théophile Brezot <theophile.brezot@cosmian.com>",
  "Bruno Grieder <bruno.grieder@cosmian.com>",
]
documentation = "https://docs.rs/cosmian_cover_crypt/"
edition = "2021"
license = "MIT/Apache-2.0"
repository = "https://github.com/Cosmian/cover_crypt"
description = "Key Policy attribute encryption based on subset cover"

[lib]
crate-type = ["rlib", "cdylib", "staticlib"]
name = "cosmian_cover_crypt"
# The cdylib is only interesting if the `--features ffi` flag is set on build
# This does not seem to be actionable conditionally https://github.com/rust-lang/cargo/issues/4881

[[bench]]
harness = false
name = "benches"

[profile.bench]
debug = true

[features]
ffi = ["lazy_static"]
full_bench = []
python = ["pyo3"]
wasm_bindgen = ["js-sys", "wasm-bindgen"]

[dependencies]
abe_policy = "1.0"
cosmian_crypto_core = { git = "https://github.com/Cosmian/crypto_core", branch = "develop" }
hex = "0.4"
js-sys = { version = "0.3", optional = true }
lazy_static = { version = "1.4", optional = true }
leb128 = "0.2"
pyo3 = { version = "0.17.3", features = ["extension-module", "abi3", "abi3-py37"], optional = true }
rand_core = { version = "0.6", features = ["getrandom"] }
serde = { version = "1.0", features = ["derive"] }
serde_json = "1.0"
sha3 = "0.10"
thiserror = "1.0"
wasm-bindgen = { version = "0.2", features = ["serde-serialize"], optional = true }
zeroize = "1.5"

[dev-dependencies]
criterion = { version = "0.4", features = ["html_reports"], default_features = false }
wasm-bindgen-test = "0.3"

```

File: pyproject.toml [hunk 16]
```diff
@@ -1,10 +1,14 @@
...
+[tool.maturin]
+# https://github.com/pypa/manylinux#manylinux
+compatibility = "manylinux_2_17"
+
 [project]
 name = "cover_crypt"
-requires-python = ">=3.6"
+requires-python = ">=3.7"
 classifiers = [
     "Programming Language :: Rust",
```
Comment (Manuthor) [inline pyproject.toml:13]: indent with 2 spaces should be enough for those 3 lines

File: src/interfaces/pyo3/py_abe_policy.rs [hunk 17]
```diff
@@ -0,0 +1,155 @@
+use abe_policy::{Attribute as AttributeRust, Policy as PolicyRust, PolicyAxis as PolicyAxisRust};
+use pyo3::{exceptions::PyException, exceptions::PyTypeError, prelude::*, types::PyType};
```
Comment (Manuthor) [inline src/interfaces/pyo3/py_abe_policy.rs:2]:
```suggestion
use pyo3::{
    exceptions::{PyException, PyTypeError},
    prelude::*,
    types::PyType,
};
```

File: src/interfaces/pyo3/py_cover_crypt.rs [hunk 18]
```diff
@@ -0,0 +1,369 @@
...
+
+use crate::{
+    api::CoverCrypt as CoverCryptRust,
+    interfaces::pyo3::py_abe_policy::Policy,
+    interfaces::statics::{
+        CoverCryptX25519Aes256, EncryptedHeader, MasterSecretKey as MasterSecretKeyRust,
+        PublicKey as PublicKeyRust, SymmetricKey as SymmetricKeyRust,
+        UserSecretKey as UserSecretKeyRust,
+    },
+};
```
Comment (Manuthor) [inline src/interfaces/pyo3/py_cover_crypt.rs:17]:
```suggestion
use pyo3::{
    exceptions::{PyException, PyTypeError},
    prelude::*,
    types::PyType,
    PyErr,
};

use crate::{
    api::CoverCrypt as CoverCryptRust,
    interfaces::{
        pyo3::py_abe_policy::Policy,
        statics::{
            CoverCryptX25519Aes256, EncryptedHeader, MasterSecretKey as MasterSecretKeyRust,
            PublicKey as PublicKeyRust, SymmetricKey as SymmetricKeyRust,
            UserSecretKey as UserSecretKeyRust,
        },
    },
};
```

File: src/interfaces/pyo3/py_cover_crypt.rs [hunk 19]
```diff
@@ -0,0 +1,369 @@
...
+}
+
+#[pyclass]
+pub struct CoverCrypt(CoverCryptX25519Aes256);
+
+#[pymethods]
+impl CoverCrypt {
+    #[new]
+    fn new() -> Self {
+        CoverCrypt(CoverCryptX25519Aes256::default())
```
Comment (Manuthor) [inline src/interfaces/pyo3/py_cover_crypt.rs:109]:
```suggestion
        Self(CoverCryptX25519Aes256::default())
```

File: src/interfaces/pyo3/tests/test.sh [hunk 20]
```diff
@@ -2,15 +2,13 @@
 
 set -euEx
 
-init(){
-  virtualenv env
-  source env/bin/activate
-  pip install maturin
-}
+# Clean previous build
```
Comment (Manuthor) [inline src/interfaces/pyo3/tests/test.sh:5]: should we finally keep this `test.sh` file? not used anywhere else?

**Review by Manuthor — COMMENTED** [review]
not much to review, great job! :+1:

File: src/interfaces/pyo3/tests/test.sh (hunk 20 above)
Comment (ackRow) [reply src/interfaces/pyo3/tests/test.sh:5]: I kept the file to quickly build and test the python package on one's local pc.

File: Cargo.toml (hunk 15 above)
Comment (ackRow) [reply Cargo.toml:47]: I think that the current version of crypto core doesn't have the right serialization implementation, @tbrezot ?

File: .gitlab-ci.yml [hunk 21]
```diff
@@ -144,6 +145,19 @@ build_android:
       - jniLibs
     expire_in: 3 mos
 
+build_python:
+  stage: build
+  image:
+    name: ghcr.io/pyo3/maturin
+    # override the image entrypoint
```
Comment (ThibaudDauce) [inline .gitlab-ci.yml:152]: It's better to explain why you need to override than explaining what you do (we can see thanks to the following line what you do)

**Pull request merged**